src/js/models/menuitem.js

```diff
@@ -86,21 +82,58 @@ define(function (require, exports, module) {
         commandKind: null,
 
         /**
-         * @type {{modifiers: number=, keyCode: number}}
+         * @type {?MenuShortcut}
          */
         shortcut: null,
 
         /**
          * @type {boolean}
          */
-        enabled: null,
+        enabled: true,
 
         /**
-         * @type {number}
+         * @type {boolean}
          */
-        checked: null
+        checked: false
     });
 
+    Object.defineProperties(MenuItem.prototype, object.cachedGetSpecs({
+        /**
+         * Maps from submenu itemID to index, excluding separators.
+         * @type {Immutable.Map.<string, number>}
+         */
+        submenuMap: function () {
+            var submenu = this.submenu;
+            if (!submenu) {
+                return null;
+            }
+
+            return Immutable.Map(submenu.reduce(function (map, entry, index) {
+                if (entry.type === "separator") {
+                    return map;
+                } else {
+                    return map.set(entry.itemID, index);
+                }
+            }, new Map()));
```

Comment:
Hey @mcilroyc, please correct me if I'm wrong, but I remember you having to change some code like this in the past because it was really slow. Am I remembering wrong?